src/fact_emitter.rs

```diff
@@ -0,0 +1,737 @@
+                                }
+                            }
+                        }
+                    }
+
+                    AccessKind::Copy | AccessKind::Move => {
+                        // FIXME: currently function call parameters are not parsed without access
+                        // kinds, check if there's some special behaviour needed for copy/moves,
+                        // instead of just being "reads" (e.g. maybe moves also need clearing
+                        // or invalidations)
+
+                        // Reads access all the origins in their type
+                        let (_, origins) = self.ty_and_origins_of_place(place);
+                        for origin in origins {
+                            facts.access_origin.push((origin.into(), node.clone()));
+                        }
+                    }
+                }
+            }
+
+            Expr::Call { arguments, .. } => {
+                // Calls evaluate their arguments
+                arguments
+                    .iter()
+                    .for_each(|expr| self.emit_expr_facts(&node, expr, facts));
+
+                // TODO: Depending on the signature of the function, some subsets can be introduced
+                // between the arguments to the call
+            }
+
+            _ => {}
+        }
+    }
+
+    // Introduce subsets: `expr` flows into `place`
+    //
+    // TODO: do we need some type checking to ensure this assigment is valid
+    // with respect to the LHS/RHS types, mutability, etc ?
+    //
+    // TODO: handles simple subsets only for now, complete this.
+    //
+    // We're in an assignment and we assume the LHS and RHS have the same shape,
+    // for example `&'a Type<&'b i32> = &'1 Type<'2 i32>`.
+    //
+    fn emit_subset_facts(&self, node: &Node, lhs_ty: &Ty, rhs_expr: &Expr, facts: &mut Facts) {
+        // Subset relationships are computed with respect to the variance rules.
+        // https://doc.rust-lang.org/reference/subtyping.html#variance
+        //
+        // In the context of an assignment, the subsets follow the flow of data, and origins on the
+        // RHS will flow into the ones on the LHS.
+        //
+        // We don't support function types in structs or function parameters at the moment, so
+        // there's no contravariant relationships yet.
+
+        let (lhs_ty, target_origin, variance) = match lhs_ty {
+            Ty::Ref {
+                origin: target_origin,
+                ty: lhs_ty,
+            } => (lhs_ty, target_origin, Variance::Covariant),
+
+            Ty::RefMut {
+                origin: target_origin,
+                ty: lhs_ty,
+            } => (lhs_ty, target_origin, Variance::Invariant),
+
+            _ => {
+                // no origins flow to the LHS
+                return;
+            }
+        };
+
+        let (rhs_ty, source_origin) = match rhs_expr {
```

Comment:
There's a similar structure, it's slightly more complicated than only relating types, since we also need to handle borrow expressions, subsets in calls, etc so this also then extracts the type from the RHS to relate it to the LHS type, in the part you mention in the comment below.